// @lazycatcloud/minidb/dist/remoteDB.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteDB = void 0;
function optionsToParams(selector, options) {
    options = Object.assign({}, options);
    // Create url
    const params = {};
    if (options.sort) {
        params.sort = JSON.stringify(options.sort);
    }
    if (options.limit != null) {
        params.limit = options.limit;
    }
    params.selector = JSON.stringify(selector || {});
    const p = new URLSearchParams(params);
    return p.toString();
}
class RemoteDB {
    constructor(url) {
        this.url = url;
    }
    async _findFetch(selector, options) {
        const params = optionsToParams(selector, options);
        return fetch(this.url + `?${params}`).then(async (res) => {
            if (!res.ok) {
                throw new Error(`HTTP error! Status: ${res.status}`);
            }
            const data = (await res.json());
            return data;
        });
    }
    find(selector, options) {
        return {
            fetch: () => this._findFetch(selector, options),
        };
    }
    async findOne(selector, options) {
        options = Object.assign({}, options, { limit: 1 });
        const params = optionsToParams(selector, options);
        return fetch(this.url + `?${params}`).then(async (res) => {
            if (!res.ok) {
                throw new Error(`HTTP error! Status: ${res.status}`);
            }
            const data = (await res.json());
            if (data.length > 0) {
                return data[0];
            }
            else {
                return null;
            }
        });
    }
    async upsertOrUpdate(selector, doc, options) {
        options = Object.assign({}, options, { limit: 1 });
        const params = optionsToParams(selector, options);
        return fetch(this.url + `/upsertOrUpdate?${params}`, {
            method: "POST",
            body: JSON.stringify(doc),
        }).then(async (res) => {
            if (!res.ok) {
                throw new Error(`HTTP error! Status: ${res.status}`);
            }
            return (await res.json());
        });
    }
    upsert(docs, bases) {
        if (!Array.isArray(docs)) {
            docs = [docs];
        }
        let request;
        if (bases) {
            if (!Array.isArray(bases)) {
                bases = Array(docs.length).fill(bases);
            }
            if (docs.length != bases.length) {
                throw "docs length don't equal bases length";
            }
            request = fetch(this.url, {
                method: "PATCH",
                body: JSON.stringify({ docs, bases }),
            });
        }
        else {
            request = fetch(this.url, {
                method: "POST",
                body: JSON.stringify({ docs }),
            });
        }
        return request.then(async (res) => {
            if (!res.ok) {
                throw new Error(`HTTP error! Status: ${res.status}`);
            }
            const data = (await res.json());
            if (docs.length > 1) {
                return data;
            }
            else {
                return data[0];
            }
        });
    }
    remove(ids) {
        if (!Array.isArray(ids)) {
            ids = [ids];
        }
        const request = fetch(this.url, {
            method: "DELETE",
            body: JSON.stringify({ ids }),
        });
        return request.then(async (res) => {
            if (!res.ok) {
                throw new Error(`HTTP error! Status: ${res.status}`);
            }
            return;
        });
    }
}
exports.RemoteDB = RemoteDB;
